### PR TITLE
GH-82448: Add thread timeout for loop.shutdown_default_executor

### LINKED ISSUE #82448: Add thread timeout for loop.shutdown_default_executor
BPO | [38267](https://bugs.python.org/issue38267)
--- | :---
Nosy | @asvetlov, @1st1, @aeros
PRs | <li>python/cpython#16360</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-09-24.19:31:44.204>
labels = ['3.9']
title = 'Add thread timeout for loop.shutdown_default_executor'
updated_at = <Date 2019-09-24.20:23:27.554>
user = 'https://github.com/aeros'
```

bugs.python.org fields:
```python
activity = <Date 2019-09-24.20:23:27.554>
actor = 'aeros'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2019-09-24.19:31:44.204>
creator = 'aeros'
dependencies = []
files = []
hgrepos = []
issue_num = 38267
keywords = ['patch']
message_count = 1.0
messages = ['353115']
nosy_count = 3.0
nosy_names = ['asvetlov', 'yselivanov', 'aeros']
pr_nums = ['16360']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue38267'
versions = ['Python 3.9']
```

</p></details>

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-82448

Co-Authored-By: Kyle Stanley <aeros167@gmail.com>
Closes https://github.com/python/cpython/issues/82448

### CODE REVIEW

File: Lib/asyncio/base_events.py [hunk 1]
```diff
@@ -561,8 +561,13 @@ async def shutdown_asyncgens(self):
                     'asyncgen': agen
                 })
 
-    async def shutdown_default_executor(self):
-        """Schedule the shutdown of the default executor."""
+    async def shutdown_default_executor(self, timeout=None):
+        """Schedule the shutdown of the default executor.
+
+        The timeout parameter specifies the amount of time the threadpool will
```
Comment (gvanrossum) [inline Lib/asyncio/base_events.py]: threadpool -> executor (also later in this docstring and again in the warning message)

File: Misc/NEWS.d/next/Library/2019-09-25-00-37-51.bpo-38267.X9Jb5V.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Add *timeout* parameter to :meth:`asyncio.loop.shutdown_default_executor`.
```
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2019-09-25-00-37-51.bpo-38267.X9Jb5V.rst:1]: Maybe give it more color? It defaults to None, but when called from the Runner it is set to 5 minutes.

File: Doc/library/asyncio-eventloop.rst [hunk 3]
```diff
@@ -178,18 +178,27 @@ Running and stopping the loop
...
+.. coroutinemethod:: loop.shutdown_default_executor(timeout=None)
 
    Schedule the closure of the default executor and wait for it to join all of
    the threads in the :class:`ThreadPoolExecutor`. After calling this method, a
    :exc:`RuntimeError` will be raised if :meth:`loop.run_in_executor` is called
    while using the default executor.
 
+   The *timeout* parameter specifies the amount of time the threadpool will
+   be given to finish joining. The default value is ``None``, which means the
+   threadpool will be given an indefinite amount of time.
```
Comment (gvanrossum) [inline Doc/library/asyncio-eventloop.rst]: Possibly cross-reference with run() and close(), where this is set to 5 minutes?

**Review by gvanrossum — COMMENTED** [review]
Nice feature!

**Comment by gvanrossum** [comment]
@kumaraditya303 Can you respond to my review, so we can land this?

File: Misc/NEWS.d/next/Library/2019-09-25-00-37-51.bpo-38267.X9Jb5V.rst (hunk 2 above)
Comment (kumaraditya303) [reply Misc/NEWS.d/next/Library/2019-09-25-00-37-51.bpo-38267.X9Jb5V.rst:1]: done

File: Lib/asyncio/base_events.py (hunk 1 above)
Comment (kumaraditya303) [reply Lib/asyncio/base_events.py]: done

File: Doc/library/asyncio-eventloop.rst (hunk 3 above)
Comment (kumaraditya303) [reply Doc/library/asyncio-eventloop.rst]: `asyncio.run` docs mention 5 minutes timeout so done.

File: Lib/asyncio/runners.py [hunk 4]
```diff
@@ -162,7 +162,7 @@ def run(main, *, debug=None):
 
     This function runs the passed coroutine, taking care of
     managing the asyncio event loop, finalizing asynchronous
-    generators and closing the threadpool.
+    generators and closing the executor.
```
Comment (gvanrossum) [inline Lib/asyncio/runners.py]:
```suggestion
    generators and closing the default executor.
```

**Review by gvanrossum — APPROVED** [review]
LGTM, I'll merge one nit and then land.

File: Misc/NEWS.d/next/Library/2019-09-25-00-37-51.bpo-38267.X9Jb5V.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+The default value is ``None``, which means the executor will be given an indefinite amount of time.
```
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2019-09-25-00-37-51.bpo-38267.X9Jb5V.rst]:
```suggestion
The default value is ``None``, which means the executor will be given an unlimited amount of time.
```

**Pull request merged**